### PR TITLE
Bugfix / Fix issues with "scan all channels"

### PR DESCRIPTION
**Pull request by itayx**
As part of the new "Scan on all channels" the client can requests a scan
with a specific pool what will indicate they wish to run a scan on all
the radio's supported channels.
A channel pool with only a single channel with a value of '0' will
indicate that the client wants to use the "Scan on all channels".

### CODE REVIEW

File: controller/src/beerocks/master/db/db.cpp [hunk 1]
```diff
@@ -2298,6 +2298,11 @@ int db::get_channel_scan_dwell_time_msec(const sMacAddr &mac, bool single_scan)
 bool db::is_channel_scan_pool_supported(const sMacAddr &mac,
                                         const std::unordered_set<uint8_t> &channel_pool)
 {
+    if (channel_pool.size() == 1 && channel_pool.find(0) != channel_pool.end()) {
```
Comment (kantera800) [inline controller/src/beerocks/master/db/db.cpp]: you should use a define 

```suggestion
    if (channel_pool.size() == 1 && *channel_pool.begin() == SCANALLCHAN) {
```

File: controller/src/beerocks/master/db/db.cpp [hunk 2]
```diff
@@ -2359,6 +2364,12 @@ bool db::is_channel_in_pool(const sMacAddr &mac, uint8_t channel, bool single_sc
 
     auto &pool =
         (single_scan ? hostap->single_scan_config : hostap->continuous_scan_config).channel_pool;
+
+    if (pool.size() == 1 && pool.find(0) != pool.end()) {
```
Comment (kantera800) [inline controller/src/beerocks/master/db/db.cpp]: same here

**Review by kantera800 — APPROVED** [review]
Approved , 
just fix the condition

File: controller/src/beerocks/master/db/db.cpp (hunk 1 above)
Comment (itayx) [reply controller/src/beerocks/master/db/db.cpp]: Changed

File: controller/src/beerocks/master/db/db.cpp (hunk 2 above)
Comment (itayx) [reply controller/src/beerocks/master/db/db.cpp]: Changed

File: controller/src/beerocks/master/db/db.cpp [hunk 3]
```diff
@@ -2298,6 +2298,11 @@ int db::get_channel_scan_dwell_time_msec(const sMacAddr &mac, bool single_scan)
 bool db::is_channel_scan_pool_supported(const sMacAddr &mac,
                                         const std::unordered_set<uint8_t> &channel_pool)
 {
+    if (channel_pool.size() == 1 && *channel_pool.begin() == SCAN_ALL_CHANNELS) {
```
Comment (tomereli) [inline controller/src/beerocks/master/db/db.cpp]: I don't like this at all. If the user wants to scan all channels, then add an API to scan on all channels. APIs should be simple - this function based on its name should only check if the given pool is supported.

File: controller/src/beerocks/master/db/db.cpp [hunk 4]
```diff
@@ -2359,6 +2364,12 @@ bool db::is_channel_in_pool(const sMacAddr &mac, uint8_t channel, bool single_sc
 
     auto &pool =
         (single_scan ? hostap->single_scan_config : hostap->continuous_scan_config).channel_pool;
+
+    if (pool.size() == 1 && *pool.begin() == SCAN_ALL_CHANNELS) {
```
Comment (tomereli) [inline controller/src/beerocks/master/db/db.cpp]: scan all channels should set the pool to empty and set a member bool scan_all_channels=true instead of this hack.

**Review by tomereli — CHANGES_REQUESTED** [review]

**Comment by itayx** [comment]
@tomereli 
I don't think that creating a new API just for the "scan on all channels" is the correct course of action,
Instead I reworked the PR to do the following:
In the `set_channel_scan_pool` I expanded the signature to accept a "scan all channels" option (set by default to `false`) which, if set to `true` will fill the DCS channel pool with the radio's supported channels.
That way the DCS task will be unaware if the client uses the "scan all channels" feature, thus containing the changes only to the DB's `set_channel_scan_pool` function and its uses in the `son_management`.
I believe this is the better course of action since it allows the DB to the the sole manager of the DCS's channel pool, and we won't need to use "hacks" to fix the issues that rise from it

File: controller/src/beerocks/master/db/db.cpp (hunk 4 above)
Comment (itayx) [reply controller/src/beerocks/master/db/db.cpp]: See my comment below

File: controller/src/beerocks/master/db/db.cpp (hunk 3 above)
Comment (itayx) [reply controller/src/beerocks/master/db/db.cpp]: See my comment below

**Review by adam1985d — COMMENTED** [review]
I think you meant to write `Preparative Commit` instead of `Preemptive commit`

**Comment by adam1985d** [comment]
> @tomereli
> I don't think that creating a new API just for the "scan on all channels" is the correct course of action,
> Instead I reworked the PR to do the following:
> In the `set_channel_scan_pool` I expanded the signature to accept a "scan all channels" option (set by default to `false`) which, if set to `true` will fill the DCS channel pool with the radio's supported channels.
> That way the DCS task will be unaware if the client uses the "scan all channels" feature, thus containing the changes only to the DB's `set_channel_scan_pool` function and its uses in the `son_management`.
> I believe this is the better course of action since it allows the DB to the the sole manager of the DCS's channel pool, and we won't need to use "hacks" to fix the issues that rise from it

I have to say I disagree with the approach taken.
I think the son_management (not the task and not the DB) should handle the operation of `get supported channels` and then call the `set_channel_scan_pool` function with the channel_pool comprised of those channels.
This way you don't change the `set_channel_scan_pool` function at all and the `son_management` is handling the user request as it should.

**Comment by itayx** [comment]
> > @tomereli
> > I don't think that creating a new API just for the "scan on all channels" is the correct course of action,
> > Instead I reworked the PR to do the following:
> > In the `set_channel_scan_pool` I expanded the signature to accept a "scan all channels" option (set by default to `false`) which, if set to `true` will fill the DCS channel pool with the radio's supported channels.
> > That way the DCS task will be unaware if the client uses the "scan all channels" feature, thus containing the changes only to the DB's `set_channel_scan_pool` function and its uses in the `son_management`.
> > I believe this is the better course of action since it allows the DB to the the sole manager of the DCS's channel pool, and we won't need to use "hacks" to fix the issues that rise from it
> 
> I have to say I disagree with the approach taken.
> I think the son_management (not the task and not the DB) should handle the operation of `get supported channels` and then call the `set_channel_scan_pool` function with the channel_pool comprised of those channels.
> This way you don't change the `set_channel_scan_pool` function at all and the `son_management` is handling the user request as it should.

I thought about it, but it will result in duplicated code for the single scan and the continuous scan.

File: common/beerocks/bcl/include/bcl/beerocks_defines.h [hunk 5]
```diff
@@ -470,6 +470,7 @@ enum class eChannelScanOpErrCode : uint8_t {
 };
 
 #define CHANNEL_SCAN_INVALID_PARAM -1
+#define SCAN_ALL_CHANNELS 0
```
Comment (tomereli) [inline common/beerocks/bcl/include/bcl/beerocks_defines.h:474]: I really don't like this (did I say this already?). I still think that splitting APIs is better, but OK, we can do this in a separate PR I guess.

File: controller/src/beerocks/master/son_management.cpp [hunk 6]
```diff
@@ -1917,7 +1917,10 @@ void son_management::handle_bml_message(Socket *sd,
             op_error_code == eChannelScanOpErrCode::CHANNEL_SCAN_OP_SUCCESS) {
             auto channel_pool_set =
                 std::unordered_set<uint8_t>(channel_pool, channel_pool + channel_pool_size);
-            op_error_code = database.set_channel_scan_pool(radio_mac, channel_pool_set, false)
+            auto scan_all_channels =
```
Comment (tomereli) [inline controller/src/beerocks/master/son_management.cpp]: So ugly...

**Review by tomereli — APPROVED** [review]
Approving cause its a bug fix, but I think this is a very bad change and we should consider splitting the API for better readability.

File: controller/src/beerocks/master/son_management.cpp (hunk 6 above)
Comment (itayx) [reply controller/src/beerocks/master/son_management.cpp]: Updated the code, and moved the logic to a helper function in the son_management

File: controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp [hunk 7]
```diff
@@ -52,54 +52,42 @@ beerocks::eChannelScanStatusCode dynamic_channel_selection_task::dcs_request_sca
...
+    request->scan_params().radio_mac         = m_radio_mac;
+    request->scan_params().dwell_time_ms     = dwell_time_msec;
+    request->scan_params().channel_pool_size = curr_channel_pool.size();
+    std::copy(curr_channel_pool.begin(), curr_channel_pool.end(),
+              request->scan_params().channel_pool);
+
     // get the parent node to send the CMDU to the agent
-    auto ire = database.get_node_parent_ire(radio_mac_str);
+    auto radio_mac_str = tlvf::mac_to_string(m_radio_mac);
+    auto ire           = database.get_node_parent_ire(radio_mac_str);
```
Comment (adam1985d) [inline controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp:94]: need to add a check here:
if we have for some reason `get_node_parent_ire` return an empty string - we don't check this and perform `send_cmdu_to_agent_ with empty std::string

File: controller/src/beerocks/master/son_management.cpp [hunk 8]
```diff
@@ -30,6 +30,34 @@ using namespace beerocks;
...
+//////////////////////////////////////////////////////////////////////////////
+
+/**
+ * @brief Set the channel pool to contain all supported channels.
+ * @param channel_pool_set:  Set containing the current channel pool.
+ * @param channel_pool_size: Size of the given set.
+ * @param database:          Reference to the DB.
+ * @param radio_mac:         MAC address of radio.
+ */
+static void set_pool_to_all_supported_channels(std::unordered_set<uint8_t> &channel_pool_set,
```
Comment (adam1985d) [inline controller/src/beerocks/master/son_management.cpp]: instead of `set_pool_to_all_supported_channels` I would suggest a `get_pool_of_all_supported_channels` which is what you do.

File: controller/src/beerocks/master/son_management.cpp [hunk 9]
```diff
@@ -30,6 +30,34 @@ using namespace beerocks;
...
+                                               uint8_t &channel_pool_size, db &database,
+                                               const sMacAddr &radio_mac)
+{
+    LOG(DEBUG) << "Setting channel pool to all channels";
+    channel_pool_set.clear();
+    auto all_channels = database.get_hostap_supported_channels(tlvf::mac_to_string(radio_mac));
```
Comment (adam1985d) [inline controller/src/beerocks/master/son_management.cpp:60]: `get_hostap_supported_channels` can fail and return an empty string `std::vector<beerocks::message::sWifiChannel>()`
please add a check here.

this means the function should return `bool` and not `void`

File: controller/src/beerocks/master/son_management.cpp [hunk 10]
```diff
@@ -1950,8 +1978,12 @@ void son_management::handle_bml_message(Socket *sd,
         }
         if (channel_pool != nullptr && channel_pool_size != CHANNEL_SCAN_INVALID_PARAM &&
             op_error_code == eChannelScanOperationCode::SUCCESS) {
+            auto u_channel_pool_size = uint8_t(channel_pool_size);
             auto channel_pool_set =
                 std::unordered_set<uint8_t>(channel_pool, channel_pool + channel_pool_size);
+            if (channel_pool_size == 1 && *channel_pool_set.begin() == SCAN_ALL_CHANNELS)
```
Comment (adam1985d) [inline controller/src/beerocks/master/son_management.cpp]: I understand why @tomereli  doesn't like the `*channel_pool_set.begin() == SCAN_ALL_CHANNELS)`
if we are going to "learn to live with it" - at least wrap it with a meaningful naming function:
```
bool is_scan_all_channels_request(channel_pool_set_type &channel_pool_set){
    return ((channel_pool_set == 1) &&*channel_pool_set.begin() == SCAN_ALL_CHANNELS));
}
```
or something similar

**Review by adam1985d — APPROVED** [review]
looks much better and cleaner now that you moved it to the son_managemet.
please fix my comments - approving

File: controller/src/beerocks/master/son_management.cpp (hunk 8 above)
Comment (itayx) [reply controller/src/beerocks/master/son_management.cpp]: Agree, much better name,
Changed

File: controller/src/beerocks/master/son_management.cpp (hunk 10 above)
Comment (itayx) [reply controller/src/beerocks/master/son_management.cpp]: The helper function actually makes the code look much better.
Changed

File: controller/src/beerocks/master/son_management.cpp (hunk 9 above)
Comment (itayx) [reply controller/src/beerocks/master/son_management.cpp:60]: Even if `get_hostap_supported_channels` fails and we get an empty `std::vector<beerocks::message::sWifiChannel>()` it won't matter since later on in the DCS task we validate the pool with
```
if (curr_channel_pool.empty()) {
        LOG(ERROR) << "empty channel pool is not supported. please set channel pool for mac="
                   << m_radio_mac;
        return beerocks::eChannelScanStatusCode::INVALID_PARAMS;
}
```
So even if an empty pool is set, the scan will fail with an `INVALID_PARAMS` in `m_last_scan_error_code`

File: controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp (hunk 7 above)
Comment (itayx) [reply controller/src/beerocks/master/tasks/dynamic_channel_selection_task.cpp:94]: Added check, in case `ire` returns as an empty string the DCS will reach the `Abort` state with an `INTERNAL_FAILURE` status

File: controller/src/beerocks/master/son_management.cpp (hunk 9 above)
Comment (adam1985d) [reply controller/src/beerocks/master/son_management.cpp:60]: we shouldn't care - the DCS task should not depend on the way we decide to implement the son_management and vise versa - or to be more precise - the dependency on the implementation should be minimal - each one should do its own part without depending on the other to verify for it.

son_management should verify it doesn't set and invalid parameter and then responds to the user "all params are ok" and the task should not send a scan request with invalid params.

**Review by adam1985d — CHANGES_REQUESTED** [review]
please see my comment on the last commit

File: controller/src/beerocks/master/son_management.cpp (hunk 9 above)
Comment (itayx) [reply controller/src/beerocks/master/son_management.cpp:60]: You're right,
Updated

File: controller/src/beerocks/master/son_management.cpp [hunk 11]
```diff
@@ -1952,6 +1994,11 @@ void son_management::handle_bml_message(Socket *sd,
             op_error_code == eChannelScanOperationCode::SUCCESS) {
             auto channel_pool_set =
                 std::unordered_set<uint8_t>(channel_pool, channel_pool + channel_pool_size);
+            if (is_scan_all_channels_request(channel_pool_set))
```
Comment (adam1985d) [inline controller/src/beerocks/master/son_management.cpp]: 1. you didn't add `{}` which is against our coding convention
2. in case of error you are still setting the DB! this part should be fixed similar to what you did in the `beerocks_message::ACTION_BML_CHANNEL_SCAN_START_SCAN_REQUEST` handling

**Review by adam1985d — CHANGES_REQUESTED** [review]

**Review by adam1985d — APPROVED** [review]
